integration-tests/tests/src/tests/observable.ts

```diff
@@ -0,0 +1,848 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2023 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+/**
+ * This suite aims to test observability of Realm, Realm.Object and all collection types.
+ * In particular, it aims to test:
+ * - Adding a listener will call the callback on change.
+ */
+
+import { expect } from "chai";
+
+import Realm, { CollectionChangeSet, DictionaryChangeSet, ObjectChangeSet } from "realm";
+
+import { openRealmBefore, openRealmBeforeEach } from "../hooks";
+import { createListenerStub } from "../utils/listener-stub";
+import { createPromiseHandle } from "../utils/promise-handle";
+
+type Observable = Realm | Realm.Object<any> | Realm.Results<any> | Realm.List<any> | Realm.Dictionary | Realm.Set<any>;
+
+function expectObservableMethods(observable: Observable) {
+  expect(observable.addListener).to.be.a("function");
+  expect(observable.removeListener).to.be.a("function");
+  expect(observable.removeAllListeners).to.be.a("function");
+}
+
+function expectRealmNotifications(
+  realm: Realm,
+  eventName: "change" | "schema" | "beforenotify",
```

Comment:
I was trying to figure out what the difference between "change" and "beforenotify" is and have realized that this isn't [documented](https://github.com/realm/realm-js/blob/main/packages/realm/src/Realm.ts#L214).  Perhaps we could add this while we are at it.

Reply:
Yeah - it's not ideal. I'll leave that for a follow up PR. To be completely honest, I don't know the intended use-case for it: Perhaps we should actually just break the API and not expose this.

Reply:
P.S. You could use [RealmEventName](https://github.com/realm/realm-js/blob/3fa8b6baf88d385e7f62546ea7b5b677eda83983/packages/realm/src/Realm.ts#L214) here, but it doesn't seem to be exported in `index.ts`. (Noticed that our [enum](https://github.com/realm/realm-js/blob/3fa8b6baf88d385e7f62546ea7b5b677eda83983/packages/realm/src/RealmListeners.ts#L21-L25) is also not in `index.ts`. So if we do expose it, we could also use those in place of the hardcoded ones.)

If I'm not mistaken, I think one of the use cases is to react to potential runtime schema change initiated on the backend? We currently simulate this in our tests via `realm._updateSchema()`.

Reply:
I created https://github.com/realm/realm-js/pull/6300.